libr/anal/sign.c

```diff
@@ -1205,75 +1236,126 @@ static bool closest_match_callback(void *a, const char *name, const char *value)
 		return false;
 	}
 
-	double score = matchGraph (it, data->test);
+	// quantify how close the signature matches
+	int div = 0;
+	double score = 0.0;
+	double gscore = -1.0;
+	if (it->graph && data->test->graph) {
+		gscore = matchGraph (it, data->test);
+		score += gscore;
+		div++;
+	}
+	double bscore = -1.0;
+	bool list_full = (r_list_length (data->output) == data->count);
+
+	// value to beat to enter the list
+	double pivot = data->score_threshold;
+	if (list_full) {
+		pivot = R_MAX (pivot, data->infimum);
+	}
+
+	if (it->bytes && data->bytes_combined) {
+		int sizea = it->bytes->size;
+		int sizeb = data->test->bytes->size;
+		if (pivot > 0.0) {
+			// bytes distance is slow. To avoid it, we can do quick maths to
+			// see if the highest possible score would be good enough to change
+			// results
+			double maxscore = R_MIN (sizea, sizeb) / R_MAX (sizea, sizeb);
+			if (div > 0) {
+				maxscore = (maxscore + score) / div;
+			}
+			if (maxscore < pivot) {
+				r_sign_item_free (it);
+				return true;
+			}
+		}
 
-	// score is too low, don't bother doing any more work
-	if (score < data->score_threshold) {
+		// get true byte score
+		bscore = cmp_bytesig_to_buff (it->bytes, data->bytes_combined, sizeb);
+		score += bscore;
+		div++;
+	}
+	if (div == 0) {
 		r_sign_item_free (it);
 		return true;
 	}
+	score /= div;
 
-	RList *output = data->output;
-
-	// return early if the list does not need to be changed
-	if (score < data->infimum && r_list_length (output) >= data->count) {
+	// score is too low, don't bother doing any more work
+	if (score < pivot) {
 		r_sign_item_free (it);
 		return true;
 	}
 
-	// remove an element if list is full
-	if (r_list_length (output) >= data->count) {
-		r_sign_close_match_free (r_list_pop (output));
-	}
-
 	// add new element
 	RSignCloseMatch *row = R_NEW (RSignCloseMatch);
 	if (!row) {
 		r_sign_item_free (it);
 		return false;
 	}
-	row->name = strdup (it->name);
-	if (!row->name) {
-		r_sign_item_free (it);
-		free (row);
-		return false;
-	}
 	row->score = score;
-	r_list_add_sorted (output, (void *)row, &score_cmpr);
+	row->gscore = gscore;
+	row->bscore = bscore;
+	row->item = it;
+	r_list_add_sorted (data->output, (void *)row, &score_cmpr);
 
-	// get new infimum
-	row = r_list_get_top (output);
-	data->infimum = row->score;
+	if (list_full) {
+		// remove smallest element
+		r_sign_close_match_free (r_list_pop (data->output));
+
+		// get new infimum
+		row = r_list_get_top (data->output);
+		data->infimum = row->score;
+	}
 
-	r_sign_item_free (it);
 	return true;
 }
 
 R_API void r_sign_close_match_free(RSignCloseMatch *match) {
-	free (match->name);
-	free (match);
+	if (match) {
+		r_sign_item_free (match->item);
+		free (match);
+	}
 }
 
 R_API RList *r_sign_find_closest_sig(RAnal *a, RSignItem *it, int count, double score_threshold) {
-	r_return_val_if_fail (a && it && count > 0 && score_threshold <= 1 && score_threshold >= 0, false);
```

Comment:
You are correct, currently `score_threshold` is always 0. I intend on adding variable in `e` to set this value to a nice default. The variable now, even though it can't be changed, to ensure the algorythm does not need a complicated change later.

I changed the `r_return_val_if_fail` to try and draw a better distinction between what is erroneous input and what is just an edge case. I guess if the caller is sending an unreasonable threshold it should probably emit a warning though. I will add score_threshold back into `r_return_val_if_fail`.